contracts/facets/ThirdPartyFacet.sol

```diff
@@ -91,117 +89,124 @@ contract ThirdPartyFacet {
 
-            if (badArch.unencryptedShardDoubleHash == shardDoubleHash) {
-                accusedArchAddresses[pos++] = matchingArchAddr;
+            // if the archaeologist has already been accused on this sarcophagus break without taking action
+            if (badArch.accused) {
+                break;
+            }
 
-                uint256 cursedBond = LibBonds.calculateCursedBond(
-                    badArch.diggingFee
-                );
+            // mark the archaeologist on the sarcophagus as having been accused
+            badArch.accused = true;
+            accusedArchAddresses[accusalCount++] = accusedArchaeologistAddress;
 
-                diggingFeesToBeDistributed += badArch.diggingFee;
-                totalCursedBond += cursedBond;
+            // track the sum of all digging fees for all accused archaeologists
+            totalDiggingFees += badArch.diggingFee;
 
-                LibBonds.decreaseCursedBond(matchingArchAddr, cursedBond);
+            // slash the accused archaeologist's bond
+            LibBonds.decreaseCursedBond(accusedArchaeologistAddress, badArch.diggingFee);
 
-                // Save the accusal against the archaeologist
-                s.archaeologistAccusals[matchingArchAddr].push(sarcoId);
-            } else {
-                revert LibErrors.AccuseIncorrectProof();
-            }
+            // Save the accusal against the archaeologist
+            s.archaeologistAccusals[accusedArchaeologistAddress].push(sarcoId);
         }
 
-        // At this point, we need to filter out unaccused archs in order to reimburse them.
-        address[] memory bondedArchaeologists = s
-            .sarcophagi[sarcoId]
-            .archaeologists;
-
-        for (uint256 i = 0; i < bondedArchaeologists.length; i++) {
-            // Need to check each archaeologist address on the sarcophagus
-            bool isUnaccused = true;
-
-            for (uint256 j = 0; j < accusedArchAddresses.length; j++) {
-                // For each arch address, if found in accusedArchAddresses,
-                // then don't add to unaccusedArchsAddresses
-                if (bondedArchaeologists[i] == accusedArchAddresses[j]) {
-                    isUnaccused = false;
-                    break;
+        // if none of the accusals were valid because the archaeologists have already been accused, return without taking action
+        if (accusalCount == 0) {
+            return;
+        }
+
+
+        address[] memory archaeologistAddresses = s.sarcophagi[sarcoId].archaeologists;
+
+        // the sarcophagus is compromised if k or more archaeologists have been accused successfully on this call
+        bool isSarcophagusCompromised = accusalCount >= sarco.minShards;
+
+        // if the current call hasn't resulted in at least k archaeologists being accused
+        // check if total number of historical accusals on sarcophagus is greater than k
+        uint historicalAccusals = 0;
```

Comment:
Could we store `historicalAccusals` in sarcoState to avoid the loop below?

Reply:
I'm thinking that it's better to have the accuser incur the cost of the loop than it is to add to the data model because accuse isn't going to be called often (and the caller is getting a payout) but an increase in the size of the Sarcophagus model impacts all users.